File: branches/new_interfaces/src/core/Algorithm.java
package core;

public class Algorithm<T extends Generation<? extends Solution>> {

	protected Initializator<T> init;
	protected Generator<T> generator;
	protected StoppingCriteria<T> stoppingCriteria;
	protected Selector<T> selector;
	protected TransitionCriteria<T> transitionCriteria;

	public Algorithm() {}
	
	public Algorithm(Initializator<T> i, Generator<T> g, StoppingCriteria<T> stop,
			Selector<T> selector, TransitionCriteria<T> trans) {

		this.init = i;
		this.generator = g;
		this.stoppingCriteria = stop;
		this.selector = selector;
		this.transitionCriteria = trans;
	}

	public T solve() {
		T currentGeneration = this.init.getInitialGeneration();

		while (!stoppingCriteria.isSatisfied(currentGeneration)) {

			T g = generator.getNext(currentGeneration);
			T h = selector.keepTheBestSolutions(g, currentGeneration);

			if (transitionCriteria.isSatisfied(currentGeneration, h)) {
				currentGeneration = h;
			}
		}

		return currentGeneration;
	}
}

File: branches/scala_implementation/taboosearch/src/core/Algorithm.scala
package core
import scala.annotation.unchecked.uncheckedVariance

class Algorithm[+S <: Solution](val initializator : Initializator[S],
                                val generator : Generator[Generation[S]],
                                val stoppingCriteria : StoppingCriteria[S],
                                val transitionCriteria : TransitionCriteria[S],
								val selector : Selector[S],
								val context : Context[S]) {
	
	def solve(initialGeneration : Generation[S] @uncheckedVariance) : Generation[S] = {
		var currentGeneration = initialGeneration;
		val evaluator = context.evaluator;
		
		while (!stoppingCriteria.isSatisfied(currentGeneration, context)) {
			var g = generator.getNext(currentGeneration);			
			var h = selector.keepTheBestSolutions(g, context);

			if (transitionCriteria.isSatisfied(currentGeneration, h)) {
				currentGeneration = h;
			}
			
			var currentSolution = currentGeneration.head;
			val currentSolutionFitness = evaluator.evaluate(currentSolution);
			
			if (currentSolutionFitness < context.bestSolutionEverFitness) {
				context.bestSolutionEver = currentSolution;
				context.bestSolutionEverFitness = currentSolutionFitness;
			}
			context.tick();
		}
		return currentGeneration;
	}
	
}